service/storage_proxy.cc

```diff
@@ -4618,6 +4659,24 @@ inet_address_vector_replica_set storage_proxy::get_live_endpoints(keyspace& ks,
     return eps;
 }
 
+inet_address_vector_replica_set storage_proxy::get_live_endpoints_for_read(keyspace& ks, const dht::token& token, db::consistency_level cl, clock_type::time_point timeout) {
+    auto& rs = ks.get_replication_strategy();
+    inet_address_vector_replica_set eps = rs.get_natural_endpoints_without_node_being_replaced(token);
+    auto itend = boost::range::remove_if(eps, [this, timeout, &gossiper = gms::get_local_gossiper()] (const gms::inet_address& ep) {
+        return !gossiper.is_alive(ep) || !is_likely_to_respond_in_time(ep, timeout);
+    });
+    // In the unlikely case when too many endpoints were disualified and it's impossible to fulfill
+    // the consistency level, bring back the endpoints which were "not likely to respond in time"
+    // and only remove the dead ones
```

Comment:
Actually, it occurs to me that the original behavior was also useful. If we think that the nodes are unlikely to respond in time, why make the attempt?

But, this requires that we have great trust in this "unlikely to respond in time".

Reply:
If the request has CL=ALL (for example) and a timeout of 10 seconds, why would you want to return an error soon instead of waiting 10 seconds and return a timeout then? You know one of the replicas is dead, but maybe it will come back in 9 seconds, and we'll be able to make the 10 second deadline? Where's the benefit of returning sooner?

The benefit we do have is when CL<ALL so speculative retry generates a lot of "background reads" which pile up without limit. So this is what this patch tries to avoid. It's not about timing out sooner for user requests.